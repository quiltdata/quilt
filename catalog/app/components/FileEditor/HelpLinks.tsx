import invariant from 'invariant'
import * as React from 'react'
import * as RRDom from 'react-router-dom'
import * as M from '@material-ui/core'
import * as Lab from '@material-ui/lab'

import Code from 'components/Code'
import Lock from 'components/Lock'
import * as quiltConfigs from 'constants/quiltConfigs'
import { docs } from 'constants/urls'
import type * as Model from 'model'
import * as BucketPreferences from 'utils/BucketPreferences'
import { createBoundary } from 'utils/ErrorBoundary'
import * as Dialogs from 'utils/GlobalDialogs'
import Log from 'utils/Logging'
import * as NamedRoutes from 'utils/NamedRoutes'
import StyledLink from 'utils/StyledLink'
import StyledTooltip from 'utils/StyledTooltip'

// TODO: put this into FileEditor/routes
function useRouteToEditFile(handle: Model.S3.S3ObjectLocation) {
  const { urls } = NamedRoutes.use()
  const { pathname, search } = RRDom.useLocation()
  const next = pathname + search
  return urls.bucketFile(handle.bucket, handle.key, { edit: true, next })
}

interface WrapperProps {
  children: React.ReactNode
}

export function WorkflowsConfigLink({ children }: WrapperProps) {
  const { bucket } = RRDom.useParams<{ bucket: string }>()
  invariant(bucket, '`bucket` must be defined')

  const toConfig = useRouteToEditFile({ bucket, key: quiltConfigs.workflows })
  return <StyledLink to={toConfig}>{children}</StyledLink>
}

const Loading = Symbol('loading')

interface AddMissingSourceBucketProps {
  bucket: string
  close: Dialogs.Close
  onSubmit: () => Promise<void>
}

function MissingSourceBucketAddConfirmation({
  bucket,
  close,
  onSubmit,
}: AddMissingSourceBucketProps) {
  const [state, setState] = React.useState<typeof Loading | Error | void>()
  const handleSubmit = React.useCallback(async () => {
    setState(Loading)
    try {
      await onSubmit()

      setState()
      close()
    } catch (error) {
      Log.error(error)
      setState(error instanceof Error ? error : new Error('Unknown error'))
    }
  }, [close, onSubmit])
  return (
    <>
      <M.DialogTitle>Add {bucket} to the source buckets list</M.DialogTitle>
      {state instanceof Error && (
        <M.DialogContent>
          <Lab.Alert severity="error">{state.message}</Lab.Alert>
        </M.DialogContent>
      )}
      {state === Loading && (
        <Lock>
          <M.CircularProgress size={32} />
        </Lock>
      )}
      <M.DialogActions>
        <M.Button onClick={close} disabled={state === Loading}>
          Cancel
        </M.Button>
        <M.Button
          color="primary"
          disabled={state === Loading}
          onClick={handleSubmit}
          variant="contained"
        >
          Update config
        </M.Button>
      </M.DialogActions>
    </>
  )
}

const DIALOG_PROPS = {
  maxWidth: 'sm' as const,
  fullWidth: true,
}

const useMissingSourceBucketTooltipStyles = M.makeStyles({
  // browsers break the word on '-'
  nowrap: {
    whiteSpace: 'nowrap',
  },
})

function MissingSourceBucketTooltip() {
  const { bucket } = RRDom.useParams<{ bucket: string }>()
  invariant(bucket, '`bucket` must be defined')

  const classes = useMissingSourceBucketTooltipStyles()

  const { handle, update } = BucketPreferences.use()

  const toConfig = useRouteToEditFile(
    handle || { bucket, key: quiltConfigs.bucketPreferences[0] },
  )

  const open = Dialogs.use()

  const autoAdd = React.useCallback(async () => {
    await update(BucketPreferences.sourceBucket(bucket))
  }, [bucket, update])

  const showConfirmation = React.useCallback(() => {
    open(
      ({ close }) => (
        <MissingSourceBucketAddConfirmation
          bucket={bucket}
          close={close}
          onSubmit={autoAdd}
        />
      ),
      DIALOG_PROPS,
    )
  }, [autoAdd, bucket, open])

  return (
    <>
      <M.Typography variant="body2" gutterBottom>
        Config property <Code>ui.sourceBuckets</Code> is empty.{' '}
        <StyledLink href={`${docs}/catalog/preferences`} target="_blank">
          Learn more
        </StyledLink>
        .
      </M.Typography>
      <M.Typography variant="body2">
        <StyledLink to={toConfig}>Edit manually</StyledLink> or{' '}
        <StyledLink onClick={showConfirmation}>
          <span className={classes.nowrap}>auto-add</span> current bucket (
          <span className={classes.nowrap}>s3://{bucket}</span>)
        </StyledLink>
      </M.Typography>
    </>
  )
}

const ErrorBoundary = createBoundary(
  (props: Lab.AlertProps) => (error: Error) => (
    <Lab.Alert severity="error" {...props}>
      {error.message || 'Unexpected Error'}
    </Lab.Alert>
  ),
  'MissingSourceBucketErrorBoundary',
)

interface MissingSourceBucketProps {
  className?: string
  children: React.ReactNode
}

const tooltipStyles = M.makeStyles((t) => ({
  tooltip: {
    maxWidth: t.spacing(36),
  },
}))

export function MissingSourceBucket({ className, children }: MissingSourceBucketProps) {
  const classes = tooltipStyles()
  return (
    <StyledTooltip
      className={className}
      classes={classes}
      interactive
      title={
        <ErrorBoundary>
          <MissingSourceBucketTooltip />
        </ErrorBoundary>
      }
    >
      <div>{children}</div>
    </StyledTooltip>
  )
}
