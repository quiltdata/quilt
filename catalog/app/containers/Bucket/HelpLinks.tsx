import * as React from 'react'
import * as RRDom from 'react-router-dom'
import * as M from '@material-ui/core'

import Skeleton from 'components/Skeleton'
import Code from 'components/Code'
import * as quiltConfigs from 'constants/quiltConfigs'
import type * as Model from 'model'
import * as BucketPreferences from 'utils/BucketPreferences'
import * as NamedRoutes from 'utils/NamedRoutes'
import StyledLink from 'utils/StyledLink'
import StyledTooltip from 'utils/StyledTooltip'

import * as requests from './requests'

// TODO: move it to components/FileEditor directory because purpose of this link is to edit file

function useRouteToEditFile(handle: Model.S3.S3ObjectLocation) {
  const { urls } = NamedRoutes.use()
  const { pathname, search } = RRDom.useLocation()
  const next = pathname + search
  return urls.bucketFile(handle.bucket, handle.key, { edit: true, next })
}

interface WrapperProps {
  children: React.ReactNode
}

export function WorkflowsConfigLink({ children }: WrapperProps) {
  const { bucket } = RRDom.useParams<{ bucket: string }>()
  const toConfig = useRouteToEditFile({ bucket, key: requests.WORKFLOWS_CONFIG_PATH })
  return <StyledLink to={toConfig}>{children}</StyledLink>
}

const useMissingSourceBucketStyles = M.makeStyles({
  nowrap: {
    whiteSpace: 'nowrap',
  },
})

interface MissingSourceBucketProps {
  className?: string
  children: React.ReactNode
}

export function MissingSourceBucket({ className, children }: MissingSourceBucketProps) {
  const { bucket } = RRDom.useParams<{ bucket: string }>()
  const classes = useMissingSourceBucketStyles()
  const { handle, update } = BucketPreferences.use()

  const toConfig = useRouteToEditFile(
    handle || { bucket, key: quiltConfigs.bucketPreferences[0] },
  )

  const [loading, setLoading] = React.useState(false)
  const handleAutoAdd = React.useCallback(async () => {
    setLoading(true)
    await update(BucketPreferences.sourceBucket(bucket))
    // Update triggers `handle` reset to `null`,
    // and `setLoading` applies to the unmounted component.
    // setLoading(false)
  }, [bucket, update])

  if (loading) return <Skeleton height={32} className={className} />

  return (
    <StyledTooltip
      className={className}
      interactive
      title={
        <>
          <M.Typography variant="body2" gutterBottom>
            Config property <Code>ui.sourceBuckets</Code> is empty.
          </M.Typography>
          <M.Typography variant="body2">
            <StyledLink to={toConfig}>Edit manually</StyledLink> or{' '}
            <StyledLink onClick={handleAutoAdd}>
              <span className={classes.nowrap}>auto-add</span> current bucket (
              <span className={classes.nowrap}>s3://{bucket}</span>)
            </StyledLink>
          </M.Typography>
        </>
      }
    >
      <div>{children}</div>
    </StyledTooltip>
  )
}
