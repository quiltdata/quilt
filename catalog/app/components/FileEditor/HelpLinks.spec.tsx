import * as React from 'react'
import renderer from 'react-test-renderer'

import { MissingSourceBucket, WorkflowsConfigLink } from './HelpLinks'

jest.mock(
  'constants/config',
  jest.fn(() => ({})),
)

jest.mock(
  'utils/StyledLink',
  () =>
    ({ href, to, children }: React.PropsWithChildren<{ href: string; to: string }>) => (
      <a href={to || href}>{children}</a>
    ),
)

jest.mock(
  'utils/StyledTooltip',
  () =>
    ({ title, children }: React.PropsWithChildren<{ title: React.ReactNode }>) => (
      <div>
        {title}
        <hr />
        {children}
      </div>
    ),
)

jest.mock('components/Code', () => ({ children }: React.PropsWithChildren<{}>) => (
  <code>{children}</code>
))

jest.mock('utils/NamedRoutes', () => ({
  ...jest.requireActual('utils/NamedRoutes'),
  use: () => ({
    urls: {
      bucketFile: (b: string, k: string, opts: Record<string, any>) => {
        const params = new URLSearchParams(opts)
        return `/b/${b}/tree/k/${k}?${params}`
      },
    },
  }),
}))

jest.mock('utils/GlobalDialogs', () => ({
  use: jest.fn(),
}))

const useLocation = jest.fn(
  () => ({ pathname: '/a/b/c', search: '?foo=bar' }) as Record<string, string>,
)

const useParams = jest.fn(() => ({ bucket: 'buck' }) as Record<string, string>)

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(() => useParams()),
  useLocation: jest.fn(() => useLocation()),
}))

describe('components/FileEditor/HelpLinks', () => {
  describe('WorkflowsConfigLink', () => {
    it('should render', () => {
      const tree = renderer
        .create(<WorkflowsConfigLink>Test</WorkflowsConfigLink>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should throw outside bucket', () => {
      jest.spyOn(console, 'error').mockImplementationOnce(jest.fn())
      useParams.mockImplementationOnce(() => ({}))
      const tree = () => renderer.create(<WorkflowsConfigLink>Any</WorkflowsConfigLink>)
      expect(tree).toThrowError('`bucket` must be defined')
    })
  })

  describe('MissingSourceBucket', () => {
    it('should render', () => {
      const tree = renderer
        .create(<MissingSourceBucket>Disabled button</MissingSourceBucket>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should throw outside bucket', () => {
      jest.spyOn(console, 'error').mockImplementationOnce(jest.fn())
      useParams.mockImplementationOnce(() => ({}))
      const tree = renderer
        .create(<MissingSourceBucket>Any</MissingSourceBucket>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
