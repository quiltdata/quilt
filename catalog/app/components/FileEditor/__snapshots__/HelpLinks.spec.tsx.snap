// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FileEditor/HelpLinks MissingSourceBucket should render 1`] = `
<div>
  <p
    className="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
  >
    Config property 
    <code>
      ui.sourceBuckets
    </code>
     is empty.
     
    <a
      href="https://docs.quiltdata.com/catalog/preferences"
    >
      Learn more
    </a>
    .
  </p>
  <p
    className="MuiTypography-root MuiTypography-body2"
  >
    <a
      href="/b/buck/tree/k/.quilt/catalog/config.yaml?edit=true&next=%2Fa%2Fb%2Fc%3Ffoo%3Dbar"
    >
      Edit manually
    </a>
     or
     
    <a>
      <span
        className="makeStyles-nowrap-2"
      >
        auto-add
      </span>
       current bucket (
      <span
        className="makeStyles-nowrap-2"
      >
        s3://
        buck
      </span>
      )
    </a>
  </p>
  <hr />
  <div>
    Disabled button
  </div>
</div>
`;

exports[`components/FileEditor/HelpLinks MissingSourceBucket should throw outside bucket 1`] = `
<div>
  <div
    className="MuiPaper-root MuiAlert-root MuiAlert-standardError MuiPaper-elevation0"
    role="alert"
  >
    <div
      className="MuiAlert-icon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
    <div
      className="MuiAlert-message"
    >
      \`bucket\` must be defined
    </div>
  </div>
  <hr />
  <div>
    Any
  </div>
</div>
`;

exports[`components/FileEditor/HelpLinks WorkflowsConfigLink should render 1`] = `
<a
  href="/b/buck/tree/k/.quilt/workflows/config.yml?edit=true&next=%2Fa%2Fb%2Fc%3Ffoo%3Dbar"
>
  Test
</a>
`;
